crates/llama_cpp/src/standard_sampler.rs

```diff
@@ -173,13 +192,42 @@ impl SamplerStage {
                 SamplerStage::TailFree(z) => {
                     llama_sample_tail_free(context, p_ptr, *z, min_keep);
                 }
+                SamplerStage::Grammar(stage) => {
+                    stage.apply(context, tokens, candidates_p, min_keep)
+                }
             }
         }
 
         candidates_p
     }
 }
 
+/// Opaque internals for [`SamplerStage::Grammar`].
+#[derive(Clone, Debug)]
+pub struct GrammarStage {
+    grammar: LlamaGrammar,
+    accepted_to: Option<usize>,
+}
+
+impl GrammarStage {
+    fn apply(
+        &mut self,
+        context: *mut llama_context,
+        tokens: &[Token],
+        mut candidates_p: llama_token_data_array,
+        _min_keep: usize,
+    ) {
+        let accepted_to = self.accepted_to.unwrap_or(tokens.len());
+        for token in &tokens[accepted_to..] {
+            unsafe { llama_grammar_accept_token(context, self.grammar.grammar.as_ptr(), token.0) }
+        }
+        self.accepted_to = Some(tokens.len());
+
+        let p_ptr = addr_of_mut!(candidates_p);
+        unsafe { llama_sample_grammar(context, p_ptr, self.grammar.grammar.as_ptr()) };
```

Comment:
I'm having a bit of a hard time following this logic but it seems like this is the implementation,

1. On the first Sampler.sample execution do not accept anything into the grammar because `accepted_to` is not set so tokens.len() will make sure the body of the for token in &tokens does not apply.
2. Set accepted_to to the current length.
3. Sample the grammar and then subsequent stages and select a token.
4. Execute the Sampler.sample again this time accepted_to is already set to the previous len of tokens so accept any new tokens into the grammar (presumably this is just 1 additional token the result of the last sample's token selection). Rinse and repeat.

Is my understanding of this correct? I'm assuming the result of the Sampler.sample is added to tokens input of the subsequent calls to sample.

Reply:
I think your understanding is correct. The `tokens` parameter is the value of `LlamaSessionInner.tokens`, which tracks the current list of tokens in the model's context, so it does update with newly generated tokens.  It is currently passed into samplers to allow for repetition penalties.

I've added a few comments and changed a few names to try to make the code clearer.